### DISCUSSION TITLE
NEL for entities that aren't already known aliases

### DISCUSSION THREAD

**Discussion by spameroo** [explosion/spaCy#7406] [Help: Coding & Implementations] [answered]
Hello, I have been attempting to train an NEL model for a task that is slightly different, but (I thought) not _that_ different from  the task in the Emerson example.  However, I'm running into issues that make me think I misunderstood something fundamental, so I'd appreciate any guidance for a way to make this work, if it's possible.

Basically, I already have an entity annotation pipeline that can label entities and link them to canonical IDs using extensive dictionaries, created from external sources of synonyms (this is working with scientific literature).  However, I also know that there are synonyms people use for these entities that _aren't_ already in our dictionaries, but which follow predictable patterns, so I have another EntityRuler component that can identify these mentions.  But I don't know what canonical entity to link them to.

My goal has been to train an EntityLinker model using all of the entities I _do_ know, and then to feed the annotations from my pattern-based EntityRuler to that EL in order to predict links for them.

What I'm seeing, though, is that any mention that did not already exist as an alias in the KnowledgeBase receives a 'NIL' prediction. (I will note that these mentions do all have vectors associated with them; the entire process has been using word vectors I trained using FastText over this same full set of literature.)

**Is the EntityLinker simply not suitable for linking mentions that don't exactly match aliases in the KB?  Or is there a way to make it do what I want, that I'm just missing?**

I can provide more detail on my configs, the KB and training example creation, etc, but first I figure I should get an answer to that first basic question.  In terms of the second, I'm wondering if maybe what I'm missing is a custom candidate generation function?

Thank you for any help!

**Comment by svlandeg** [comment, marked as answer]
Hi @spameroo, apologies for the late follow-up. Since spaCy v.3, you can implement your own custom candidate generation step. More details are here: https://github.com/explosion/spaCy/discussions/7801#discussioncomment-666505

The default one indeed only looks for exact matches, and feeds those results into the actual entity linker that does the disambiguation step based on the vectors. You could also use the vectors in your custom candidate generation step and do some sort of "nearest neigbour" search, though I imagine the performance of that will heavily depend on the size of your Knowledge base.
